File: 0198-house-robber/0198-house-robber.java
class Solution {

    public int f(int i, int nums[],int dp[]){
        if(i<0) return 0;
        if(i==0) return nums[i];
        if(dp[i]!=-1) return dp[i];
        int pick = nums[i]+ f(i-2,nums,dp);
        int notpick = f(i-1,nums,dp);
        return dp[i] = Math.max(pick,notpick);
    }
    public int rob(int[] nums) {
        int n= nums.length;
        int [] dp =new int[n];
        Arrays.fill(dp,-1);
        int ans=f(n-1,nums,dp);
        return ans;
    }
}

File: 0198-house-robber/0198-house-robber.c
int rob(int* nums, int numsSize) 
{
    int temp,temp0,n=numsSize,t;
    if(n==1) return nums[0];
    if(n==2) return fmax(nums[0],nums[1]);
    temp=nums[0];
    temp0=fmax(nums[0],nums[1]);
    for(int i=2;i<n;i++)
    {
        t=fmax(temp+nums[i],temp0);
        temp=temp0;
        temp0=t;
    }
    return temp0;

}